crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -2228,6 +2255,104 @@ where
     }
 }
 
+impl SemanticSyntaxContext for SemanticIndexBuilder<'_> {
+    fn seen_docstring_boundary(&self) -> bool {
+        self.seen_module_docstring_boundary
+    }
+
+    fn future_annotations_or_stub(&self) -> bool {
+        self.has_future_annotations
+    }
+
+    fn python_version(&self) -> PythonVersion {
+        Program::get(self.db).python_version(self.db)
+    }
+
+    fn with_source<T>(&self, f: impl FnOnce(&str) -> T) -> T {
+        let src = source_text(self.db.upcast(), self.file);
+        f(src.as_str())
+    }
+
+    // TODO(brent) handle looking up `global` bindings
+    fn global(&self, _name: &str) -> Option<TextRange> {
+        None
+    }
+
+    fn in_async_context(&self) -> bool {
+        for scope_info in self.scope_stack.iter().rev() {
+            let scope = &self.scopes[scope_info.file_scope_id];
+            match scope.kind() {
+                ScopeKind::Class | ScopeKind::Lambda => return false,
+                ScopeKind::Function => {
+                    if let Some(f) = scope.node().as_function() {
+                        return f.is_async;
+                    }
+                }
+                ScopeKind::Comprehension
+                | ScopeKind::Module
+                | ScopeKind::TypeAlias
+                | ScopeKind::Annotation => {}
+            }
+        }
+        false
+    }
+
+    fn in_await_allowed_context(&self) -> bool {
+        for scope_info in self.scope_stack.iter().rev() {
+            let scope = &self.scopes[scope_info.file_scope_id];
+            match scope.kind() {
+                ScopeKind::Class => return false,
+                ScopeKind::Function | ScopeKind::Lambda => return true,
+                ScopeKind::Comprehension
+                | ScopeKind::Module
+                | ScopeKind::TypeAlias
+                | ScopeKind::Annotation => {}
+            }
+        }
+        false
+    }
+
+    fn in_sync_comprehension(&self) -> bool {
+        for scope_info in self.scope_stack.iter().rev() {
+            let scope = &self.scopes[scope_info.file_scope_id];
+            let generators = match scope.node() {
+                NodeWithScopeKind::ListComprehension(node) => &node.generators,
+                NodeWithScopeKind::SetComprehension(node) => &node.generators,
+                NodeWithScopeKind::DictComprehension(node) => &node.generators,
+                _ => continue,
+            };
+            if generators.iter().all(|gen| !gen.is_async) {
+                return true;
+            }
+        }
+        false
+    }
+
+    fn in_module_scope(&self) -> bool {
+        self.scopes[self.current_scope()].kind().is_module()
```

Comment:
we already have a method for this one ;)

```suggestion
        self.current_scope_is_global_scope()
```

but maybe we should just delete that method, move the implementation of that method to the trait method here, and refactor the callsites of the method to use the trait method?

```suggestion
        self.scope_stack.len() == 1
```

Reply:
I figured I would miss at least one existing method, thanks!

Hmm, I was worried about having to import the trait everywhere this method is used, but so far it's only used in this file. I'll move the implementation to the trait then!